File: leetcode/python/P0040_CombinationSum2.py
# https://leetcode.com/problems/combination-sum-ii
from typing import List


class Solution:
    def combinationSum2(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []

        def findCombinations(combination: List[int], target_sum: int, idx: int) -> None:
            if target_sum == 0:
                result.append(combination)
                return
            if target_sum < 0:
                return

            for i in range(idx, len(candidates)):
                # sorted the candidates array, the values only increase
                if target_sum - candidates[i] < 0:
                    break
                if i > idx and candidates[i] == candidates[i - 1]:
                    continue
                findCombinations(
                    combination + [candidates[i]], target_sum - candidates[i], i + 1
                )

        candidates.sort()
        findCombinations([], target, 0)
        return result


if __name__ == "__main__":
    solution = Solution()

    for ii, testcase in enumerate((([10, 1, 2, 7, 6, 1, 5], 8),)):
        candidates, target = testcase
        print(f"Testcase: {ii + 1} {candidates} {target}]\n")
        print(f"Solution: {solution.combinationSum2(candidates, target)}\n")

File: leetcode/cpp/P0040_CombinationSum2.cpp
/* https://leetcode.com/problems/combination-sum-ii */
#include <algorithm>
#include <iostream>
#include <vector>

class Solution {
 public:
  std::vector<std::vector<int>> combinationSum2(
      const std::vector<int>& candidates, int target) {
    std::vector<std::vector<int>> result;
    if (candidates.size() == 0) {
      return {{}};
    }
    std::vector<int> sorted_candidates(candidates);
    std::sort(sorted_candidates.begin(), sorted_candidates.end());

    std::vector<int> mask(candidates.size(), 0);
    findCombinations(sorted_candidates, target, 0, mask, result);
    return result;
  }

  inline void findCombinations(const std::vector<int>& candidates, int target,
                               int index, std::vector<int>& mask,
                               std::vector<std::vector<int>>& results) {
    if (target < 0) {
      return;
    }
    if (target == 0) {
      std::vector<int> sol;
      for (int jj = 0; jj < candidates.size(); ++jj) {
        if (mask[jj] == 1) {
          sol.push_back(candidates[jj]);
        }
      }

      results.emplace_back(sol);
    }

    for (int ii = index; ii < candidates.size(); ++ii) {
      if (target - candidates[ii] < 0) {
        break;
      }
      if (ii > index && candidates[ii] == candidates[ii - 1]) {
        continue;
      }
      if (mask[ii] == 0) {
        mask[ii] = 1;
        findCombinations(candidates, target - candidates[ii], ii + 1, mask,
                         results);
        mask[ii] = 0;
      }
    }
  }
};

void print_vec(const std::vector<int>& vec) {
  for (int x : vec) {
    std::cout << x << " ";
  }
  std::cout << "\n";
}

int main(int argc, char** argv) {
  Solution s;
  std::vector<std::vector<int>> result =
      s.combinationSum2({10, 1, 2, 7, 6, 1, 5}, 8);
  for (const auto& vec : result) {
    print_vec(vec);
  }
}